packages/devtools_app/lib/src/charts/chart_controller.dart

```diff
@@ -0,0 +1,449 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:intl/intl.dart' as intl;
+
+import '../auto_dispose.dart';
+
+import 'chart_trace.dart';
+
+/// Zoom level of chart
+enum displayDuration { live, fiveMinutes, fifteenMinutes, all }
+
+/// Given a durationLabel return the displayDuration.
+displayDuration durationToDisplayDuration(
+  displayDuration currentZoomDuration,
+  Duration duration,
+) {
+  if (currentZoomDuration == displayDuration.live) {
+    if (duration.inMinutes == 0 && duration.inSeconds >= 10) {
+      return displayDuration.live;
+    }
+  } else if (currentZoomDuration == displayDuration.fifteenMinutes) {
+    if (duration.inMinutes >= 5) {
+      return displayDuration.fifteenMinutes;
+    }
+  } else if (currentZoomDuration == displayDuration.fiveMinutes) {
+    if (duration.inMinutes == 3) {
+      return displayDuration.fiveMinutes;
+    }
+  }
+
+  return null;
+}
+
+String prettyTimestamp(int timestamp) {
+  final timestampDT = DateTime.fromMillisecondsSinceEpoch(timestamp);
+  return intl.DateFormat.Hms().format(timestampDT); // HH:mm:ss
+}
+
+/// Indexes into xAxisLabeledTimestamps.
+const leftLabelIndex = 0;
+const centerLabelIndex = 1;
+const rightLabelIndex = 2;
+
+/*
+                                          xPaddingRight
+                                          |
+      leftPadding          topPadding     |  rightPadding
+      |                    |              |  |
+    ==|====================|==============|==|=
+    " |                    V              |  |
+    " V ----------------------------------V  V
+    "   |
+    "   |
+    "   |
+    "   | ------ canvasChartWidth --------
+    "   |
+    "   |
+    "   ---------------------------------- <----- yCanvasChart
+    "   '  '  '  '  '  '  '  '  '  '  '  '    {-------- bottomPadding
+    "   ^                ^                     \
+    ====|================|======================
+        |                |
+        xCanvasChart     X-axis ticks
+
+*/
+class ChartController extends DisposableController
+    with AutoDisposeControllerMixin {
+  ChartController({
+    this.displayXLabels = true,
+    this.displayYLabels = true,
+  });
+
+  // Total size of area to display the chart (includes title, axis, labels, etc.).
+  Size size;
+
+  // TODO(terry): Compute dynamically based on title text height.
+  final topPadding = 20.0;
+  // TODO(terry): Compute dynamically based on Y-axis lables text width.
+  final leftPadding = 50.0;
+  // Computed minimum right padding.
+  int rightPadding = 25;
+  // TODO(terry): Compute dynamically based on X-axis lables text height.
+  final bottomPadding = 40.0;
+
+  // TODO(terry): Consider computation of each X-axis tick width (scaled).
+  double tickWidth = 10.0;
```

Comment:
should these be private with getters?

Reply:
got idea tickWidth should be private with getter.

Done.